test/statement_types.py

```diff
@@ -0,0 +1,451 @@
+
+class RngVariable(CppStatement):
+    """Represents a random number generator"""
+    def __init__(self, name):
+        """
+        Initialize an rng object
+        
+        :param name: C++ name of variable
+        """
+        self.name = name
+    
+    def cpp(self):
+        """Generate C++"""
+        return "std::minstd_rand {name};".format(name = self.name)
+
+class OStreamVariable(CppStatement):
+    """Represents an ostream pointer"""
+    def __init__(self, name):
+        """
+        Initialize an ostream pointer
+        
+        :param name: C++ name of variable
+        """
+        self.name = name
+    
+    def cpp(self):
+        """Generate C++"""
+        return "std::ostream* {name} = &std::cout;".format(name = self.name)
+
+class ArrayVariable(CppStatement):
+    """Represents an array variable"""
+    def __init__(self, overload, name, number_nested_arrays, inner_type, value = None, size = None):
+        """
+        Initialize an array
+        
+        :param overload: Type of overload as string (Prim/Fwd/Rev/etc.)
+        :param name: C++ name of variable
+        :param number_nested_arrays: Number of array dimensions
+        :param inner_type: Stanc3 type string of inner type of array
+        :param value: Either a scalar value (represented as another CppStatement) to initialize the array with or an iterable type of Python values (not CppStatements)
+        :param size: Size of arrays
+        """
+        self.overload = overload
+        self.name = name
+        self.number_nested_arrays = number_nested_arrays
+        self.inner_type = inner_type
+        self.value = value
+
+        if not size:
+            self.size = 1
+        else:
+            self.size = size
+    
+    def is_reverse_mode(self):
+        """
+        If the underlying value is a CppStatement, return the reverse-mode-ness of that underlying variable
+
+        Otherwise return true if the overload is reverse mode and inner type isn't an integer
+        """
+        try:
+            return self.value.is_reverse_mode()
+        except AttributeError:
+            return self.overload.startswith("Rev") and self.inner_type != "int"
+
+    def is_varmat_compatible(self):
+        """Return true if the underlying value is a CppStatement and is varmat compatible"""
+        try:
+            return self.value.is_varmat_compatible()
+        except AttributeError:
+            return False
+    
+    def cpp(self):
+        """Generate C++"""
+        try:
+            lhs_string = "decltype({name})".format(name = self.value.name)
+            rhs_string = self.value.name
+            for n in range(self.number_nested_arrays):
+                lhs_string = "std::vector<" + lhs_string + ">"
+                rhs_string = "{" + ",".join([rhs_string] * self.size) + "}"
+
+            return "{lhs_string} {name} = {rhs_string};".format(lhs_string = lhs_string, name = self.name, rhs_string = rhs_string)
```

Comment:
Try to avoid long try blocks. The way this is written I have no idea which statement you expect could throw. (unless almost any statment could, but I think this is not the case here)

Reply:
I changed the try to an if statement to make the condition clearer.